File: November/8/minsizesubarrsum.py
# LeetCode 209: Minimum Size Subarray Sum
# Explanation:
# 1. Use two pointers (start, end) to maintain a sliding window.
# 2. Expand end to increase sum, contract start to minimize length.
# 3. Track minimal window where sum >= target.
# Time Complexity: O(n)
# Space Complexity: O(1)

import bisect

class Solution:
    def minSubArrayLen(self, target: int, nums: list[int]) -> int:
        prefix = [0]
        for n in nums:
            prefix.append(prefix[-1] + n)
        res = float('inf')
        for i in range(len(nums)):
            desired = prefix[i] + target
            j = bisect.bisect_left(prefix, desired)
            if j < len(prefix):
                res = min(res, j - i)
        return 0 if res == float('inf') else res

File: November/8/minsizesubarrsum.js
// LeetCode 209: Minimum Size Subarray Sum
// Explanation:
// 1. Maintain a sliding window using two pointers.
// 2. Expand right until sum >= target, then shrink from left.
// 3. Track the smallest window length found.
// Time Complexity: O(n)
// Space Complexity: O(1)

var minSubArrayLen = function(target, nums) {
    let left = 0, sum = 0, minLen = Infinity;
    for (let right = 0; right < nums.length; right++) {
        sum += nums[right];
        while (sum >= target) {
            minLen = Math.min(minLen, right - left + 1);
            sum -= nums[left++];
        }
    }
    return minLen === Infinity ? 0 : minLen;
};
